opt: index join col stat not implemented

I am working on a change that tries to derive logical properties from all expressions in the memo, and cross-check them against the group properties. This found a crash related to missing implementation of the col stats for index join. It can be reproed with this test:

```
opt colstat=1 colstat=2 colstat=3 colstat=4 colstat=(1,2,3)
SELECT * FROM a WHERE s = 'foo' AND x + y = 10
----
```

```
panic during testdata/stats/index-join:81:
SELECT * FROM a WHERE s = 'foo' AND x + y = 10
panic: unrecognized relational expression type: index-join [recovered]
	panic: unrecognized relational expression type: index-join [recovered]
	panic: unrecognized relational expression type: index-join

goroutine 56 [running]:
testing.tRunner.func1(0xc000028f00)
	/usr/local/go/src/testing/testing.go:792 +0x387
panic(0x10bc3e0, 0xc00027f150)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/testutils/datadriven.RunTest.func1.1(0xc000673620)
	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/datadriven/datadriven.go:75 +0x117
panic(0x10bc3e0, 0xc00027f150)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*statisticsBuilder).colStat(0xc0005f18e8, 0x2, 0x0, 0xc0001fb190, 0xd00000010, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/statistics_builder.go:307 +0x82e
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*statisticsBuilder).colStatFromChild(0xc0005f18e8, 0x2, 0x0, 0xc0001fb190, 0x50000000c, 0x0, 0x10)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/statistics_builder.go:190 +0x12f
```

Note that I found and fixed another similar crash for lookup join. These suggest we don't have good coverage on the col stat code. Perhaps we want an test option like `colstat` but which triggers column statistics (e.g. single-column statistics for all output cols) for all operators (not just the top one).

--- a/pkg/sql/opt/memo/statistics_builder.go
+++ b/pkg/sql/opt/memo/statistics_builder.go
@@ -530,7 +530,13 @@ func (sb *statisticsBuilder) colStatSelect(colSet opt.ColSet, ev ExprView) *prop
 	s := &relProps.Stats
 	inputStats := &ev.childGroup(0).logical.Relational.Stats
 	colStat := sb.copyColStatFromChild(colSet, ev, s)
-	colStat.ApplySelectivity(s.Selectivity, inputStats.RowCount)
+
+	// It's not safe to use s.Selectivity, because it's possible that some of the
+	// filter conditions were pushed down into the input after s.Selectivity
+	// was calculated. For example, an index scan or index join created during
+	// exploration could absorb some of the filter conditions.
+	selectivity := s.RowCount / inputStats.RowCount
+	colStat.ApplySelectivity(selectivity, inputStats.RowCount)
 	return colStat
 }
 
@@ -850,9 +856,44 @@ func (sb *statisticsBuilder) colStatIndexJoin(
 	relProps := ev.Logical().Relational
 	s := &relProps.Stats
 
-	// TODO(#30288): implement this.
+	def := ev.Private().(*IndexJoinDef)
+	inputProps := ev.Child(0).Logical().Relational
+	inputCols := inputProps.OutputCols
+
 	colStat, _ := s.ColStats.Add(colSet)
 	colStat.DistinctCount = 1
+
+	// Some of the requested columns may be from the input index.
+	reqInputCols := colSet.Intersection(inputCols)
+	if !reqInputCols.Empty() {
+		inputColStat := sb.colStatFromChild(reqInputCols, ev, 0 /* childIdx */)
+		colStat.DistinctCount = inputColStat.DistinctCount
+	}
+
+	// Other requested columns may be from the primary index.
+	reqLookupCols := colSet.Difference(inputCols).Intersection(def.Cols)
+	if !reqLookupCols.Empty() {
+		// Make a copy of the lookup column stats so we don't modify the originals.
+		lookupColStat := *sb.colStatTable(def.Table, reqLookupCols)
+
+		// Calculate the distinct count of the lookup columns given the selectivity
+		// of any filters on the input.
+		inputStats := &inputProps.Stats
+		tableStats := sb.makeTableStatistics(def.Table)
+		selectivity := inputStats.RowCount / tableStats.RowCount
+		lookupColStat.ApplySelectivity(selectivity, tableStats.RowCount)
+
+		// Multiply the distinct counts in case colStat.DistinctCount is
+		// already populated with a statistic from the subset of columns
+		// provided by the input index. Multiplying the counts gives a worst-case
+		// estimate of the joint distinct count.
+		colStat.DistinctCount *= lookupColStat.DistinctCount
+	}
+
+	// The distinct count should be no larger than the row count.
+	if colStat.DistinctCount > s.RowCount {
+		colStat.DistinctCount = s.RowCount
+	}
 	return colStat
 }
 

--- a/pkg/sql/opt/memo/testdata/stats/index-join
+++ b/pkg/sql/opt/memo/testdata/stats/index-join
@@ -42,8 +42,8 @@ ALTER TABLE a INJECT STATISTICS '[
 ]'
 ----
 
-# In order to actually create new logical props for the lookup join, we
-# need to call ConstructLookupJoin, which only happens when where is a
+# In order to actually create new logical props for the index join, we
+# need to call ConstructIndexJoin, which only happens when there is a
 # remaining filter.
 opt
 SELECT count(*) FROM (SELECT * FROM a WHERE s = 'foo' AND x + y = 10) GROUP BY s, y
@@ -77,3 +77,76 @@ project
       │         └── (x + y) = 10 [type=bool, outer=(1,2)]
       └── aggregations
            └── count-rows [type=int]
+
+opt colstat=1 colstat=2 colstat=3 colstat=4 colstat=(1,2,3)
+SELECT * FROM a WHERE s = 'foo' AND x + y = 10
+----
+select
+ ├── columns: x:1(int!null) y:2(int) s:3(string!null) d:4(decimal!null)
+ ├── stats: [rows=66.6666667, distinct(1)=66.6666667, distinct(2)=52.9461341, distinct(3)=1, distinct(4)=60.3661899, distinct(1-3)=66.6666667]
+ ├── key: (1)
+ ├── fd: ()-->(3), (1)-->(2,4), (4)-->(1,2)
+ ├── index-join a
+ │    ├── columns: x:1(int!null) y:2(int) s:3(string) d:4(decimal!null)
+ │    ├── stats: [rows=200, distinct(1)=200, distinct(2)=87.8423345, distinct(4)=130.264312, distinct(1-3)=200]
+ │    ├── key: (1)
+ │    ├── fd: ()-->(3), (1)-->(2,4), (4)-->(1), (3,4)~~>(1,2)
+ │    └── scan a@secondary
+ │         ├── columns: x:1(int!null) s:3(string!null) d:4(decimal!null)
+ │         ├── constraint: /-3/4: [/'foo' - /'foo']
+ │         ├── stats: [rows=200, distinct(1)=200, distinct(3)=1, distinct(4)=130.264312, distinct(1,3)=200]
+ │         ├── key: (1)
+ │         └── fd: ()-->(3), (1)-->(4), (4)-->(1)
+ └── filters [type=bool, outer=(1,2)]
+      └── (x + y) = 10 [type=bool, outer=(1,2)]
+
+opt colstat=1 colstat=2 colstat=3 colstat=(1,2,3)
+SELECT * FROM a WHERE s = 'foo'
+----
+index-join a
+ ├── columns: x:1(int!null) y:2(int) s:3(string!null) d:4(decimal!null)
+ ├── stats: [rows=200, distinct(1)=200, distinct(2)=87.8423345, distinct(3)=1, distinct(1-3)=200]
+ ├── key: (1)
+ ├── fd: ()-->(3), (1)-->(2,4), (4)-->(1,2)
+ └── scan a@secondary
+      ├── columns: x:1(int!null) s:3(string!null) d:4(decimal!null)
+      ├── constraint: /-3/4: [/'foo' - /'foo']
+      ├── stats: [rows=200, distinct(1)=200, distinct(3)=1, distinct(1,3)=200]
+      ├── key: (1)
+      └── fd: ()-->(3), (1)-->(4), (4)-->(1)
+
+# Note that the row count of the index join does not match the row count of
+# the scan, because the index join's row count was carried over from the
+# normalized SELECT expression in its memo group (see next test case).
+# In order to fix the row count, we need more precise constraint calculation
+# for filters.
+opt colstat=1 colstat=2 colstat=3 colstat=(2,3) colstat=(1,2,3)
+SELECT * FROM a WHERE s = 'foo' OR s = 'bar'
+----
+index-join a
+ ├── columns: x:1(int!null) y:2(int) s:3(string) d:4(decimal!null)
+ ├── stats: [rows=666.666667, distinct(1)=400, distinct(2)=98.8470785, distinct(3)=2, distinct(2,3)=197.694157, distinct(1-3)=666.666667]
+ ├── key: (1)
+ ├── fd: (1)-->(2-4), (3,4)~~>(1,2)
+ └── scan a@secondary
+      ├── columns: x:1(int!null) s:3(string!null) d:4(decimal!null)
+      ├── constraint: /-3/4: [/'foo' - /'foo'] [/'bar' - /'bar']
+      ├── stats: [rows=400, distinct(1)=400, distinct(3)=2, distinct(1,3)=400]
+      ├── key: (1)
+      └── fd: (1)-->(3,4), (3,4)-->(1)
+
+norm
+SELECT * FROM a WHERE s = 'foo' OR s = 'bar'
+----
+select
+ ├── columns: x:1(int!null) y:2(int) s:3(string) d:4(decimal!null)
+ ├── stats: [rows=666.666667]
+ ├── key: (1)
+ ├── fd: (1)-->(2-4), (3,4)~~>(1,2)
+ ├── scan a
+ │    ├── columns: x:1(int!null) y:2(int) s:3(string) d:4(decimal!null)
+ │    ├── stats: [rows=2000]
+ │    ├── key: (1)
+ │    └── fd: (1)-->(2-4), (3,4)~~>(1,2)
+ └── filters [type=bool, outer=(3)]
+      └── (s = 'foo') OR (s = 'bar') [type=bool, outer=(3)]

